Merge CoreMangLib, Exceptions and Interop folders

trylek opened the pull request:
This change converts tests in the CoreMangLib, Exceptions and Interop to the merged model. As next step I plan to work on review feedback to my previous PR merging baseservices and on merging the remaining tests.

Thanks

Tomas

trylek committed:
Fix visibilities and return values per analyzer issues

trylek committed:
Mechanically convert CoreMangLib tests to merged model

trylek committed:
Convert Exceptions tests to merged model

trylek committed:
Small test fixes (unneeded return values, visibilities, Assert variants)

trylek committed:
Additional individual test fixes

trylek committed:
Remaining individual test project changes

trylek committed:
Replace explicit references to CoreCLRTestLibrary with $(TestLibraryProjectPath)

trylek committed:
Mechanically merge the Interop folder

trylek committed:
Fix UnmanagedCallersOnlyTest by making the negative case a separate test

trylek committed:
Revert incorrect changes to two tests

trylek commented:
/cc @ivsiazsa @dotnet/runtime-infrastructure

markples commented on src/tests/CoreMangLib/system/delegate/VSD/OpenDelegate.csproj:
```diff
@@ -1,5 +1,7 @@
   <PropertyGroup>
+    <!-- Needed for mechanical merging of all remaining tests, this particular project may not actually need process isolation -->
+    <RequiresProcessIsolation>true</RequiresProcessIsolation>
     <OutputType>exe</OutputType>
```
It looks like the lowercase 'exe' slipped through the scan.

markples replied:
(though, now that I think about it, I guess that's where ReqProcIso ends up anyway, though I think we automatically set it (kind of hacky in Dir.B.targets since it is after the SDK has had a look at it))

trylek replied:
I think you're right that we shouldn't keep obsolete ```OutputType=exe``` properties on most tests except the tiny subset opting out of the merged wrappers for special reasons like the half a dozen tests under baseservices that do weird things. I have experimentally removed all these annotations from the interop tests and I'm retesting the change locally, I'll be happy to update the PR by removing those if it turns out to be working.

markples commented on src/tests/Interop/MonoAPI/MonoMono/InstallEHCallback.cs:
```diff
@@ -5,183 +5,183 @@
+        MonoAPI.Tests.MonoAPISupport.Setup();
+        int result;
+        result = test_0_setjmp_exn_handler ();
+        Assert.Equal(0, result);
```
I suspect that this (and the others) end up being fine (mono test, int equality), but I don't know for sure that these don't end up being run in some configuration that breaks.  But hopefully outerloop (and extra platforms?) is good enough to check that?

trylek replied:
All I can say to this now is that in general Mono tests are supported by the merged test infrastructure, @jkoritzinsky went to great lengths to make that fully functional and IIRC Mono people were happy about his implementation in terms of efficiency and cleanness.

markples replied:
Sorry, I wasn't clear.  I just don't know if this test will end up running in some other configuration (from a quick look, I don't see it being excluded from other runtimes).  However, I meant it more as a "heads up" in case something breaks, not an objection to the PR.

jkoritzinsky replied:
If we have concerns, we can use the SkipOnCoreClrAttribute on the test and it will automatically be skipped on non-Mono platforms. Tbh I think that's a better solution than suppressing the test in issues.targets.

markples commented on src/tests/Interop/PInvoke/Array/MarshalArrayAsParam/AsDefault/AsDefaultTest.cs:
```diff
@@ -395,7 +395,7 @@ private static void TestMarshalInOut_ByVal()
         int[] iArrLength0 = InitArray<int>(0);
         Assert.True(CStyle_Array_Int_InOut_ZeroLength(iArrLength0));
-        Assert.Equal(0, iArrLength0.Length);
+        Assert.Empty(iArrLength0);
```
similarly, I have no idea whether int equality or `Empty` is more dangerous in the various configurations

trylek replied:
I just fixed those as the analyzer slapped me because of them and it seemed to me that somehow exempting them from the scanning would be more complicated than just going with the flow.

markples commented on src/tests/Interop/PInvoke/BestFitMapping/Program.cs:
```diff
@@ -6,37 +6,33 @@
-            LPStr.PInvoke_False_True.RunTest();
-            LPStr.PInvoke_True_False.RunTest();
-            LPStr.PInvoke_True_True.RunTest();
-            return 100;
-        } catch (Exception e){
-            Console.WriteLine($"Test Failure: {e}");
-            return 101;
-        }
```
I like this cleanup that we get!

trylek committed:
Remove obsolete OutputType=exe properties per Mark's PR feedback

markples commented on src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.cs:
```diff
@@ -45,11 +46,6 @@ public static int Main(string[] args)
                 TestUnmanagedCallersOnlyViaUnmanagedCalli_ThrowException();
             }
-
-            if (args.Length != 0 && args[0].Equals("calli"))
```
Was this being tested before?  (if not, you'll get a +1 in your counts)

jkoritzinsky replied:
No, this was only for manual testing. In the interop space, we do very little automated testing of "kills the process" behaviors around invalid code patterns.

trylek replied:
The test count is a complete mess, under the merged scheme the test count is more than twice the original number. I think this is because multiple tests under this subtree used the ```[Fact]``` annotations even before test merging and in the legacy system those reported as a single test within a particular project whereas today they smoothly blend with the overall infra and report as multiple tests. There were two tests exercising CoreCLR runtime crashes under special circumstances, the bit you have pointed out and a similar bit in the SuppressGCTransition test. Apparently they haven't been running recently - we may have broken them at some point in the past by removing the RunOnly projects (tests running other tests with specific sets of command-line parameters) or perhaps they were always broken. So now they are broken no more and I expect follow-up work once they start throwing different errors on other platforms.

trylek replied:
OK, thanks @jkoritzinsky, so please let me know whether I should keep these two as negative tests or throw them away.

jkoritzinsky replied:
I think it's worth keeping them as Pri1 tests, at least for now.

trylek committed:
Fix exit code for the negative tests and make them Pri1 per Jeremy's PR feedback

trylek committed:
Manually convert 6 Interop tests overlooked by ILTransform

trylek committed:
Fix the Interop/MonoAPI/MonoMono/Thunks test

trylek committed:
Mark out of process tests with Private="False"

This is needed so that they don't copy their outputs to the merged
wrapper folder, triggering errors in Mono builds because some output
file names overlap.

Thanks

Tomas

trylek committed:
Fix a new test, block copying custom System.Private.CoreLib to output

trylek commented:
/azp run runtime-coreclr outerloop

trylek commented:
/azp run runtime-extra-platforms
